predicates/primitive/shutdown.go

```diff
@@ -0,0 +1,51 @@
+package primitive
+
+import (
+	"net/http"
+	"os"
+	"os/signal"
+	"sync/atomic"
+	"syscall"
+
+	"github.com/zalando/skipper/predicates"
+	"github.com/zalando/skipper/routing"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type shutdown struct {
+	inShutdown int32
+}
+
+// NewShutdown provides a predicate spec to create predicates
+// that evaluate to true if Skipper is shutting down
+func NewShutdown() routing.PredicateSpec {
+	s, _ := newShutdown()
+	return s
+}
+
+func newShutdown() (routing.PredicateSpec, chan os.Signal) {
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGTERM)
+	s := &shutdown{}
+	go func() {
+		<-sigs
+		log.Infof("Got shutdown signal for %s predicates", s.Name())
+		atomic.StoreInt32(&s.inShutdown, 1)
+	}()
+	return s, sigs
+}
+
+func (*shutdown) Name() string { return predicates.ShutdownName }
+
+// Create returns a Predicate that evaluates to true if Skipper is shutting down
+func (s *shutdown) Create(args []interface{}) (routing.Predicate, error) {
+	if len(args) != 0 {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+	return s, nil
+}
+
+func (s *shutdown) Match(*http.Request) bool {
+	return atomic.LoadInt32(&s.inShutdown) != 0
+}
```

Comment:
@szuecs 
```
$ go tool objdump bin/skipper | grep -F 'primitive.(*shutdown).Match' -A5
TEXT github.com/zalando/skipper/predicates/primitive.(*shutdown).Match(SB) /.../shutdown.go
  shutdown.go:50        0xf6d3c0                8b08                    MOVL 0(AX), CX          
  shutdown.go:50        0xf6d3c2                85c9                    TESTL CX, CX            
  shutdown.go:50        0xf6d3c4                0f95c0                  SETNE AL                
  shutdown.go:50        0xf6d3c7                c3                      RET      
```

Reply:
```
  % cat foo.go
package main

import (
        "fmt"
        "time"
)

var i int

func foo() bool {
        return i != 0 // line 11
}

func main() {
        foo()
        go func() {
                i = 1  // line 17
        }()
        for i := 0; i < 100; i++ {
                if foo() {  // line 20
                        println("not 0")
                }
                time.Sleep(1)
        }
        fmt.Println("end")
}
```

To show that all of it is atomic:
```
% go tool objdump foo | grep -E 'foo.go:(11|17|20)'
  foo.go:11             0x47e48b                48833d2d690d0000        CMPQ $0x0, main.i(SB)
  foo.go:20             0x47e493                74d5                    JE 0x47e46a
  foo.go:17             0x47e520                48c70595680d0001000000  MOVQ $0x1, main.i(SB)
```

Not sure if this is atomic enough, but I will ask in Go slack because now I am interested in this :)

Reply:
@AlexanderYastrebov there was an interesting answer that in x86 (the same seems true for amd64) sync.Load is implemented for int/int32/int64 https://go.godbolt.org/z/WsToKnxrs as a single MOVQ instruction so the result seems to be consistent if we use it or not and should not have any perf benefits if we drop it.
So let's go with your current code which is of course correct. :)